File: src/main/java/io/github/romeoahmed/cursedoath/world/SweptVolume.java
package io.github.romeoahmed.cursedoath.world;

import net.minecraft.world.phys.AABB;
import net.minecraft.world.phys.Vec3;
import org.jspecify.annotations.Nullable;

/// A sphere sweep or conservative oriented box sweep; bounds are only the broad phase.
public final class SweptVolume {
    private final Vec3 start, end, halfSize, forward, side, up, destination;
    private final @Nullable SphereSweep sphere;
    private final AABB bounds;

    public SweptVolume(Vec3 start, Vec3 end, Vec3 halfSize) {
        this(start, end, halfSize, false);
    }

    public SweptVolume(Vec3 start, Vec3 end, Vec3 halfSize, boolean rounded) {
        this.start = start;
        this.end = end;
        this.halfSize = halfSize;
        var direction = end.subtract(start).normalize();
        forward = direction.equals(Vec3.ZERO) ? new Vec3(0, 0, 1) : direction;
        var reference = Math.abs(forward.y) < 0.99 ? new Vec3(0, 1, 0) : new Vec3(0, 0, 1);
        side = forward.cross(reference).normalize();
        up = side.cross(forward).normalize();
        destination = new Vec3(0, 0, end.distanceTo(start));
        sphere = rounded ? new SphereSweep(start, end, halfSize.x) : null;
        bounds = rounded
                ? new AABB(start, end).inflate(halfSize.x)
                : new AABB(start, end)
                        .inflate(
                                Math.abs(side.x) * halfSize.x
                                        + Math.abs(up.x) * halfSize.y
                                        + Math.abs(forward.x) * halfSize.z,
                                Math.abs(side.y) * halfSize.x
                                        + Math.abs(up.y) * halfSize.y
                                        + Math.abs(forward.y) * halfSize.z,
                                Math.abs(side.z) * halfSize.x
                                        + Math.abs(up.z) * halfSize.y
                                        + Math.abs(forward.z) * halfSize.z);
    }

    public Vec3 start() {
        return start;
    }

    public Vec3 end() {
        return end;
    }

    public Vec3 forward() {
        return forward;
    }

    public AABB bounds() {
        return bounds;
    }

    public @Nullable Double entry(AABB box) {
        return entry(box, Vec3.ZERO);
    }

    public @Nullable Double entry(AABB box, Vec3 movement) {
        if (sphere != null)
            return (movement.equals(Vec3.ZERO) ? sphere : new SphereSweep(start.add(movement), end, halfSize.x))
                    .entry(box);
        var center = local(box.getCenter().subtract(start));
        var half = new Vec3(box.getXsize() / 2, box.getYsize() / 2, box.getZsize() / 2);
        var from = local(movement);
        var extent = new Vec3(projectedSize(side, half), projectedSize(up, half), projectedSize(forward, half))
                .add(halfSize);
        var localBox = new AABB(center.subtract(extent), center.add(extent));
        if (localBox.contains(from)) return 0.0;
        var hit = localBox.clip(from, destination).orElse(null);
        return hit == null ? null : hit.distanceTo(from) / from.distanceTo(destination);
    }

    public @Nullable Vec3 contact(AABB box) {
        return contact(box, Vec3.ZERO);
    }

    public @Nullable Vec3 contact(AABB box, Vec3 movement) {
        var time = entry(box, movement);
        if (time == null) return null;
        var point = start.lerp(end, time).add(movement.scale(1 - time));
        return new Vec3(
                Math.clamp(point.x, box.minX, box.maxX),
                Math.clamp(point.y, box.minY, box.maxY),
                Math.clamp(point.z, box.minZ, box.maxZ));
    }

    public Vec3 source(Vec3 point) {
        return source(point, start);
    }
    /// Projects a cover-ray origin just behind the attack's original rear plane.
    ///
    /// @param point destination whose transverse position is preserved
    /// @param origin original attack center, retained across advancing segments
    /// @return ray origin aligned with the attack direction
    public Vec3 source(Vec3 point, Vec3 origin) {
        return point.subtract(forward.scale(point.subtract(origin).dot(forward) + halfSize.z + 0.001));
    }

    private Vec3 local(Vec3 vector) {
        return new Vec3(vector.dot(side), vector.dot(up), vector.dot(forward));
    }

    private static double projectedSize(Vec3 axis, Vec3 half) {
        return Math.abs(axis.x) * half.x + Math.abs(axis.y) * half.y + Math.abs(axis.z) * half.z;
    }
}

File: src/main/kotlin/io/github/romeoahmed/cursedoath/world/SweptVolume.kt
package io.github.romeoahmed.cursedoath.world

import net.minecraft.world.phys.AABB
import net.minecraft.world.phys.Vec3
import kotlin.math.abs

/** A sphere sweep or conservative oriented box sweep; bounds are only the broad phase. */
internal class SweptVolume(
    val start: Vec3,
    val end: Vec3,
    private val halfSize: Vec3,
    rounded: Boolean = false,
) {
    val forward = end.subtract(start).normalize()
    private val reference = if (abs(forward.y) < VERTICAL_THRESHOLD) Vec3(0.0, 1.0, 0.0) else Vec3(0.0, 0.0, 1.0)
    private val side = forward.cross(reference).normalize()
    private val up = side.cross(forward).normalize()
    private val destination = Vec3(0.0, 0.0, end.distanceTo(start))
    private val sphere = if (rounded) SphereSweep(start, end, halfSize.x) else null
    val bounds: AABB =
        if (rounded) {
            AABB(start, end).inflate(halfSize.x)
        } else {
            AABB(start, end).inflate(
                abs(side.x) * halfSize.x + abs(up.x) * halfSize.y + abs(forward.x) * halfSize.z,
                abs(side.y) * halfSize.x + abs(up.y) * halfSize.y + abs(forward.y) * halfSize.z,
                abs(side.z) * halfSize.x + abs(up.z) * halfSize.y + abs(forward.z) * halfSize.z,
            )
        }

    fun entry(
        box: AABB,
        movement: Vec3 = Vec3.ZERO,
    ): Double? {
        if (sphere != null) {
            val sweep = if (movement == Vec3.ZERO) sphere else SphereSweep(start.add(movement), end, halfSize.x)
            return sweep.entry(box)
        }
        val center = local(box.center.subtract(start))
        val half = Vec3(box.xsize / 2, box.ysize / 2, box.zsize / 2)
        val motion = local(movement)
        val extent =
            Vec3(
                projectedSize(side, half),
                projectedSize(up, half),
                projectedSize(forward, half),
            ).add(halfSize)
        val localBox = AABB(center.subtract(extent), center.add(extent))
        val from = motion
        return if (localBox.contains(from)) {
            0.0
        } else {
            localBox.clip(from, destination).orElse(null)?.let {
                it.distanceTo(from) /
                    from.distanceTo(destination)
            }
        }
    }

    fun contact(
        box: AABB,
        movement: Vec3 = Vec3.ZERO,
    ): Vec3? {
        val time = entry(box, movement) ?: return null
        val point = start.lerp(end, time).add(movement.scale(1 - time))
        return Vec3(
            point.x.coerceIn(box.minX, box.maxX),
            point.y.coerceIn(box.minY, box.maxY),
            point.z.coerceIn(box.minZ, box.maxZ),
        )
    }

    /** Parallel cover rays start behind the original attack, never inside a wall it has passed. */
    fun source(
        point: Vec3,
        origin: Vec3 = start,
    ): Vec3 = point.subtract(forward.scale(point.subtract(origin).dot(forward) + halfSize.z + SURFACE_MARGIN))

    private fun local(vector: Vec3) = Vec3(vector.dot(side), vector.dot(up), vector.dot(forward))

    private fun projectedSize(
        axis: Vec3,
        half: Vec3,
    ): Double = abs(axis.x) * half.x + abs(axis.y) * half.y + abs(axis.z) * half.z

    private companion object {
        const val VERTICAL_THRESHOLD = 0.99
        const val SURFACE_MARGIN = 0.001
    }
}
